Fix scrolling

H3rnand3zzz opened the pull request:
Current state:
 - Fixes "overscrolling"
 - Lag reduction (only necessary lines are being drawn)

Currently it's full of debugging lines with some issues which are to be fixed before release:
- [x] Fix "underscrolling"
- [x] Fix "overscrolling" (profanity jumps over big messages received from DB)
- [x] Change line calculation (made based on curses' position)
- [x] Cleanup debug lines
- [x] Merge
- [x] Fix a problem: extremely big message (e.g. 3k lines) causes profanity to jump back once scrolled on top

#### This line is for github CI/CD system to link the issue
Fix https://github.com/profanity-im/profanity/issues/1934

H3rnand3zzz commented:
### New revision

 - Git commit description is removed for now since a lot of things are subject to change.
 - New approach: let ncurses calculate lines instead of trying to do it by ourselves. Passing numbers of lines in the buffer helped to resolve overscrolling problem and can be handy in the future (jumping to some messages (unread, search), visual selection of messages etc.).
 - Underscrolling is broken again since some numbers are off in line calculation (though they are mostly correct, it's likely caused by some edge cases), requires further investigation.

H3rnand3zzz commented:
### New revision
 - Commit message that outlines changes
 - Merged with the latest changes
 - Added fetching `stanza_id` from the DB (changes are highly interconnected)
 - Removed debug messages

H3rnand3zzz commented:
### New revision
 - Comments are removed as per @sjaeckel's request.

jubalh commented:
> Pass id from DB (I am not sure if it's correct, probably stanza-id should be passed, but also we should consider LMC case)

What were your thoughts here exactly? I think id that you used is correct.

jubalh commented on src/ui/window.c:
```diff
@@ -656,15 +654,20 @@ win_page_up(ProfWin* window, int scroll_size)
         }
     }
 
-    // went past beginning, show first page
```
why is this comment removed? I think its helpful to quickly see whats going on here.

jubalh commented on src/ui/window.c:
```diff
@@ -656,15 +654,20 @@ win_page_up(ProfWin* window, int scroll_size)
+    }
 
     window->layout->paged = 1;
-    // update only if position has changed
```
same as above

H3rnand3zzz commented:
### New revision
Addressed the issues outlined in the review.

jubalh commented on src/ui/window.c:
```diff
@@ -715,7 +723,7 @@ win_page_down(ProfWin* window, int scroll_size)
         *page_start = total_rows - page_space - 1;
 
     window->layout->paged = 1;
-    // update only if position has changed
```
same

H3rnand3zzz commented:
addressed

jubalh commented:
>  Fix a problem: extremely big message (e.g. 3k lines) causes profanity to jump back once scrolled on top

Was this already the case before your change?
I'm not even sure 3k lines fit on a screen? :D

H3rnand3zzz commented:
### New revision
 - Removed buffer limitation based on amount of lines in messages: it might affect users who don't use history, the approach has a good idea, but requires more careful implementation (e.g. limitation on amount of lines drawn as opposed to stored lines)
 - Description updated
 - Rebased on master

jubalh commented:
>  it might affect users who don't use history, the approach has a good idea, but requires more careful implementation

Affect how? Should we wait with merging/reviewing the code before this is solved? I would prefer to not include something "wrong" in master because several of our users use master to help us find bugs. So this might annoy them or make them report something that we are aware of.

H3rnand3zzz commented:
> > it might affect users who don't use history, the approach has a good idea, but requires more careful implementation
> 
> Affect how? Should we wait with merging/reviewing the code before this is solved? I would prefer to not include something "wrong" in master because several of our users use master to help us find bugs. So this might annoy them or make them report something that we are aware of.

It might've affected in a previous version, currently it was removed, hence no one should be affected in any negative way (at least I don't see a problem currently, I might've missed something obviously).

jubalh commented on src/database.c:
```diff
@@ -365,8 +365,10 @@ log_database_get_previous_chat(const gchar* const contact_barejid, const char* s
+        char* id = (char*)sqlite3_column_text(stmt, 6);
 
         ProfMessage* msg = message_init();
+        msg->id = id ? strdup(id) : NULL;
```
I see you introduced `STRDUP_OR_NULL` in `buffer.c`. I wonder if it would be worth it to have this in common.h and use it here (and in other places) as well? @sjaeckel ?

jubalh replied:
@sjaeckel ping

jubalh commented on src/ui/buffer.c:
```diff
@@ -50,23 +50,30 @@
 #include "ui/buffer.h"
 
-#define BUFF_SIZE 200
+#define BUFF_SIZE           200
```
Could you add a comment for `BUFF_SIZE` and `BUFF_SIZE_LINES` to short explain what they stand for?

H3rnand3zzz replied:
I renamed the variable to self-describing name.

H3rnand3zzz commented:
### New revision
 - Fixed the issue pointed out above (buffer wasn't limited at all)
